x/auth/keeper/address_space.go

```diff
@@ -0,0 +1,98 @@
+package keeper
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+
+	"cosmossdk.io/collections"
+
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+)
+
+// AddressSpaceManager is an interface for managing foreign address spaces.
+type AddressSpaceManager interface {
+	// Name is the name of the address space.
+	Name() string
+
+	// DeriveAddress derives an address from the given account ID and public key.
+	// Implementations should attempt to follow the rules for their address space
+	// when the public key is of a known type (ex. secp256k1) and should perform
+	// a suitable derivation for other key types when the public key is unknown or nil.
+	// The account ID passed to this method will always be non-nil, but the
+	// public key may be nil.
+	// Implementations should not panic if the public key is nil and should instead
+	// perform derivation simply based on the account ID.
+	DeriveAddress(id AccountID, pubKey cryptotypes.PubKey) Address
+}
```

Comment:
would it be possible to get a simple example impl of this that works for the basic cosmos address stuff we have now?